### PR TITLE
Server SHOULD send full-sized datagrams until the path is validated

### LINKED ISSUE #4183: Handshake failure (infinite ping-pong) when path MTU is asymmetric
Consider the following scenario, assuming that the MTU from the client to the server is 1200 bytes, and the MTU from the server to client is 1000 bytes.
* step 1. Client sends:
  * datagram (1200 bytes), carrying Initial (Hello).
* step 2. In response, the server sends:
  * datagram (1200 bytes), carrying Initial (Hello) and a Handshake packet that contains the first half of handshake-level TLS messages (i.e. EncryptedExtensions, Certificate, CertificateVerify, Finished).
  * datagram (500 bytes), carrying Initial (PING) and a Handshake packet that contains the latter half of the handshake-level TLS messages.

The first datagram gets lost, the second datagram reaches the client.

* step 3. In response to the datagram being received, the client sends:
  * datagram (tiny), carrying Initial (ACK).
* step 4. When the server receives the packet, it learns the RTT, and arms the time-based loss detection timer. In a fraction of RTT, that timer fires, and the server sends what have been sent in step 2.

The first of the two datagrams sent in step 4 gets lost, the second one reaches the client.

* step 5. In response to the datagram being received, the client sends an ACK, as it did in step 3.

As we can see, we have an infinite loop that cannot make progress.

This example might sound a little strange, as the server arms the time-based loss detection timer for Handshake packets when receiving an ACK for the Initial packet. Though I would point out that the specification is lax on how an endpoint might use an Initial ACK to drive loss recovery of other packet number spaces; see [the third paragraph of section 6.2.3 of the recovery draft](https://quicwg.org/base-drafts/draft-ietf-quic-recovery.html#section-6.2.3-3).

We might pretend that the MAY in [the second paragraph of section 6.2.3](https://quicwg.org/base-drafts/draft-ietf-quic-recovery.html#section-6.2.3-2) restricts alternative loss signals (i.e. duplicate CRYPTO frames, receipt of Handshake packets before obtaining Handshake keys) from being applied to anything other than Initial PN space. But even in that case, I think we still have a failure mode. If the size of ServerHello is slightly above the MTU, if the server is designed to embed ACK in the first Initial packet for each send event, and if the client skips packet numbers upon observing possible loss, the size of the server-sent ACK frame increases two for every Initial packet that the client sends. This means that for each round-trip, the space that can be used for sending CRYPTO data will decrease by 2 bytes for the first Initial packets that the server sends in each iteration. These 2 bytes will be sent as tiny packets, and they will reach the client, while the full-size packet gets consistently dropped. In other words, the failure mode in this case is that there would be a ping-pong between the endpoints, and only 2 bytes of data can be exchanged per each round-trip.

### PR DESCRIPTION
**Pull request by kazuho**
Closes #4183.

The recommendation is written so that it would prevent the two failure modes discussed in the issue:
* client and server doing ping-pong due to only the tail of the server-sent packets reaching the client
* the connection getting stuck after the handshake is complete, when Initial and Handshake packets will be small (particularly when TLS session tickets are used)

### CODE REVIEW

**Review by martinthomson — COMMENTED** [review]
I think that this is better moved to [Section 14.1](https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#name-initial-packet-size), with a reference back to this section instead.  The purpose of this section is to talk about address validation, not the minimum MTU guarantee, which is what this recommendation safeguards.

File: draft-ietf-quic-transport.md [hunk 1]
```diff
@@ -1943,6 +1943,11 @@ the client during connection establishment with a Retry packet (see
 {{validate-retry}}) or in a previous connection using the NEW_TOKEN frame (see
 {{validate-future}}).
 
+Until the path is validated, a server SHOULD pad the Initial, Handshake, 1-RTT
+packets it sends so that their UDP payload size will be at least 1200 bytes.
+Doing so guarantees that the handshake would make progress only when the path is
+capable of handling QUIC traffic (see {{packet-size}}).
```
[the remark this answers is not in the document]
Comment (martinthomson) [reply draft-ietf-quic-transport.md:1949]: Maybe also:

```suggestion
capable of handling QUIC traffic in both directions; see {{packet-size}}.
```

File: draft-ietf-quic-transport.md [hunk 2]
```diff
@@ -1943,6 +1943,11 @@ the client during connection establishment with a Retry packet (see
 {{validate-retry}}) or in a previous connection using the NEW_TOKEN frame (see
 {{validate-future}}).
 
+Until the path is validated, a server SHOULD pad the Initial, Handshake, 1-RTT
+packets it sends so that their UDP payload size will be at least 1200 bytes.
```
[the remark this answers is not in the document]
Comment (kazuho) [reply draft-ietf-quic-transport.md:1947]: The server cannot pad Retry packets.

File: draft-ietf-quic-transport.md [hunk 3]
```diff
@@ -1943,6 +1943,11 @@ the client during connection establishment with a Retry packet (see
 {{validate-retry}}) or in a previous connection using the NEW_TOKEN frame (see
 {{validate-future}}).
 
+Until the path is validated, a server SHOULD pad the Initial, Handshake, 1-RTT
```
Comment (marten-seemann) [inline draft-ietf-quic-transport.md:1946]: You need to pad Handshake packets, if you consistently pad Initials: Without the ServerHello, the client won't be able to decrypt the Handshake packet.

**Review by marten-seemann — CHANGES_REQUESTED** [review]

File: draft-ietf-quic-transport.md (hunk 3 above)
Comment (kazuho) [reply draft-ietf-quic-transport.md:1946]: This is necessary to prevent the ping-pong that is being described in [the issue](https://github.com/quicwg/base-drafts/issues/4183#issue-715343219) which is caused by the receipt of a Handshake packet triggering a retransmission.

**Comment by kazuho** [comment]
@martinthomson Thank you for the review.

> I think that this is better moved to Section 14.1, with a reference back to this section instead. The purpose of this section is to talk about address validation, not the minimum MTU guarantee, which is what this recommendation safeguards.

Is the suggestion to move also the client-side padding requirement in section 8.1 (MUST pad Initial packets) to section 14.1? The intent of adding the text to section 8.1 was so that the paragraph goes like:
* client MUST pad all Initial packets
* servers MAY can Retry
* servers SHOULD pad packets until the path is being validated.

I'm fine with expanding the last bullet point to cover both endpoints, though I thin we might want to keep the first bullet point and the third point in the same section?

File: draft-ietf-quic-transport.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (martinthomson) [reply draft-ietf-quic-transport.md:1947]:
```suggestion
Until the path is validated, a server SHOULD pad all packets it sends, other
than Retry, so that their UDP payload size will be at least 1200 bytes.
```

**Review by martinthomson — COMMENTED** [review]
I see your logic about placement and that makes sense.  I would ask that you reference this from 14.1 though, as that is where people might look for packet size requirements.

Thanks for pointing out Retry; all except Retry would be fine.

File: draft-ietf-quic-transport.md (hunk 3 above)
Comment (marten-seemann) [reply draft-ietf-quic-transport.md]: If you pad all Initial packets (meaning: all datagrams containing Initial), you don't need to pad Handshake packets. An undecryptable packet doesn't reset any timers, so if won't lead to the ping-pong you describe.

This is easier to implement (for me at least), since I can use the same logic for packing packets on the client and server side: Pad all packets that contain an Initial.

Comment (nibanks) [reply draft-ietf-quic-transport.md]: I agree with @marten-seemann. If the first flight pads in both directions, ACKs of those packets should be enough to validate the 1200 MTU requirement, and further packets shouldn't be required to pad, right?

**Review by janaiyengar — APPROVED** [review]
Editorial suggestion, but this LGTM, thanks @kazuho for writing it up.

@marten-seemann @nibanks : https://quicwg.org/base-drafts/draft-ietf-quic-recovery.html#section-6.2.3-1 suggests that a client can retransmit on receiving HS packets before it has keys for them, which can cause a ping-pong.

File: draft-ietf-quic-transport.md (hunk 3 above)
Comment (kazuho) [reply draft-ietf-quic-transport.md]: @marten-seemann @nibanks I'm not sure if I agree.

Let me start by pointing out the basics that I think is true, before I dive into the details of the recovery draft:
* When the path MTU is below 1200 bytes, it is better to make sure that none of the packets reach the client rather than letting some pass through, because client might extend their timeout if they receive some packets from the server.
* Machinery defined in the transport draft should not depend on the specifics of the recovery draft.
* What we are trying to provide is a recommendation (SHOULD), and that should be the best practice rather than the minimum that an endpoint is supposed to do. And because it is merely a recommendation, endpoints can adopt alternative strategies.

Regarding the specifics of the recovery draft, [the second paragraph of the recovery draft](https://quicwg.org/base-drafts/draft-ietf-quic-recovery.html#section-6.2.3-2) states that an server MAY retransmit CRYPTO data before the PTO timer expires, when it receives an Initial packet containing duplicate CRYPTO data.

The question is if this paragraph prohibits servers from retransmitting Handshake packets (or 1-RTT packets) containing CRYPTO data. If such action is prohibited, then we will not have ping-pong. If such action is _not_ prohibited, we might see ping-ping if we only pad datagrams carrying Initial packets. That is because if only the Handshake packets (or the 1-RTT packets) reach the client, the client is explicitly allowed to retransmit Initial packets containing CRYPTO data.

My understanding is that such action is _not_ prohibited. In fact, I would argue that the intent of this section is to allow such an action.

Consider the case where the server sent three datagrams in its first flight: 1) a full-sized datagram carrying an Initial and a Handshake packet, 2) a full-sized datagram carrying a Handshake packet, 3) a half-sized datagram carrying a Handshake packet.

When the server receives an Initial carrying duplicate CRYPTO data, the draft states that it can assume that the Initial that the server sent has been lost. What does that exactly mean? The natural assumption would be that either or both of datagrams _2_ and _3_ reached the client, but _1_ did not. It is far less likely that the client received all the three datagrams, but there was a bit flip in the Initial packet, and that the client is buffering the Handshake packets it was unable to decrypt.

Then, why should a server limit the retransmission to only Initial data?

To summarize, my understanding is that section 6.2.3 of the recovery draft suggests that the server MAY retransmit Handshake packets when receiving an Initial packet that contain duplicate CRYPTO data. Therefore, there's the possibility that we would see a ping-pong even with a server that is following the recommendations in the recovery draft, unless it pads all packets that it sends until the path is validated.

**Comment by kazuho** [comment]
@martinthomson Thank you for the re-review, back reference has been added in 8ec73e1.

File: draft-ietf-quic-transport.md [hunk 4]
```diff
@@ -4099,8 +4104,9 @@ bytes.  A server MAY also immediately close the connection by sending a
 CONNECTION_CLOSE frame with an error code of PROTOCOL_VIOLATION; see
 {{immediate-close-hs}}.
 
-The server MUST also limit the number of bytes it sends before validating the
-address of the client; see {{address-validation}}.
+Until the server validates the client address, a server MUST also limit the
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md:4107]:
```suggestion
Until the server validates the client address, it MUST limit the
```

File: draft-ietf-quic-transport.md [hunk 5]
```diff
@@ -4099,8 +4104,9 @@ bytes.  A server MAY also immediately close the connection by sending a
...
+the number of bytes it sends and SHOULD pad the Initial packets; see
+{{address-validation}}.
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md:4109]:
```suggestion
number of bytes it sends, and it SHOULD ensure that the UDP
payload of all datagrams it sends at least 1200 bytes; see
{{address-validation}}.
```

File: draft-ietf-quic-recovery.md [hunk 6]
```diff
@@ -663,9 +663,10 @@ in Initial packets, or the client's estimated RTT is too small. When a
 client receives Handshake or 1-RTT packets prior to obtaining Handshake keys,
 it may assume some or all of the server's Initial packets were lost.
 
-To speed up handshake completion under these conditions, an endpoint MAY send
-a packet containing unacknowledged CRYPTO data earlier than the PTO expiry,
-subject to the address validation limits in Section 8.1 of {{QUIC-TRANSPORT}}.
+To speed up handshake completion under these conditions, an endpoint MAY, for a
+limited number of occasions per each connection, send a packet containing
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md:667]: Why is this necessary?

File: draft-ietf-quic-transport.md [hunk 7]
```diff
@@ -1943,6 +1943,11 @@ the client during connection establishment with a Retry packet (see
 {{validate-retry}}) or in a previous connection using the NEW_TOKEN frame (see
 {{validate-future}}).
 
+Until the client address is validated, servers SHOULD ensure that UDP datagrams
+containing Initial packets have UDP payloads of at least 1200 bytes.  Doing so
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md:1947]: I would revert this to what you had earlier... I don't think there's a good reason to not recommend padding all packets, since client behavior is uncertain here. The SHOULD here means that implementations can deviate from this recommendation, as long as they know what they're doing.

**Review by janaiyengar — CHANGES_REQUESTED** [review]
Huh, it looks like you've changed your recommendation to only Initial packets. I think that ought to be all packets.

File: draft-ietf-quic-transport.md (hunk 7 above)
Comment (marten-seemann) [reply draft-ietf-quic-transport.md:1947]: Retry can validate the client's address, so the ping-pong could still happen in that case.

Comment (kazuho) [reply draft-ietf-quic-transport.md]: @marten-seemann Wow. That's true. Addressed in bd089c9.

File: draft-ietf-quic-recovery.md (hunk 6 above)
Comment (martinthomson) [reply draft-ietf-quic-recovery.md:667]: This is the optimization that leads to the infinite regress.  This is attempting to say "A client MUST limit the number of times it sends Initial packets in response to packets it cannot process."

File: draft-ietf-quic-transport.md (hunk 5 above)
Comment (martinthomson) [reply draft-ietf-quic-transport.md:4106]: I want to keep the Initial piece.

File: draft-ietf-quic-transport.md [hunk 8]
```diff
@@ -4099,8 +4101,9 @@ bytes.  A server MAY also immediately close the connection by sending a
 CONNECTION_CLOSE frame with an error code of PROTOCOL_VIOLATION; see
 {{immediate-close-hs}}.
 
-The server MUST also limit the number of bytes it sends before validating the
-address of the client; see {{address-validation}}.
+Until the server validates the client address, it MUST limit the
+the number of bytes it sends and SHOULD pad the Initial packets; see
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md:4105]:
```suggestion
the number of bytes it sends and SHOULD pad Initial packets; see
```

File: draft-ietf-quic-transport.md [hunk 9]
```diff
@@ -1920,10 +1920,12 @@ that are uniquely attributed to a single connection. This includes datagrams
 that contain packets that are successfully processed and datagrams that contain
 packets that are all discarded.
 
-Clients MUST ensure that UDP datagrams containing Initial packets have UDP
-payloads of at least 1200 bytes, adding PADDING frames as necessary.
-A client that sends padded datagrams allows the server to
-send more data prior to completing address validation.
+Clients MUST and servers SHOULD ensure that UDP datagrams containing Initial
+packets have UDP payloads of at least 1200 bytes, adding PADDING frames as
+necessary.  A client that sends padded datagrams allows the server to send more
```
Comment (kazuho) [inline draft-ietf-quic-transport.md:1925]: Maybe it would be better to limit the scope and say _servers SHOULD pad Initial packets that carry crypto data_.

The problem of padding all Initial packets is that ACK-only Initial packets consume full MTU, when the send window is still controlled by the amplification limit.

File: draft-ietf-quic-recovery.md (hunk 6 above)
Comment (ianswett) [reply draft-ietf-quic-recovery.md:667]: FWIW, I am happy with this being limited to once.

File: draft-ietf-quic-transport.md (hunk 9 above)
Comment (ianswett) [reply draft-ietf-quic-transport.md:1925]: I think making a carve-out for ACK-only packets sounds attractive, but it could result in the client sending an Initial, getting it acknowledged, but not receiving the server's Initial.  Then things get complex to reason about.

Comment (ianswett) [inline draft-ietf-quic-transport.md:1925]: @martinduke suggestion:

```suggestion
Any datagram that contains an Initial packet MUST be at least 1200 Bytes unless the packet contains
a CONNECTION_CLOSE frame.  A client that sends padded datagrams allows the server to send more
```

File: draft-ietf-quic-transport.md [hunk 10]
```diff
@@ -4099,8 +4101,9 @@ bytes.  A server MAY also immediately close the connection by sending a
 CONNECTION_CLOSE frame with an error code of PROTOCOL_VIOLATION; see
 {{immediate-close-hs}}.
 
-The server MUST also limit the number of bytes it sends before validating the
-address of the client; see {{address-validation}}.
+Until the server validates the client address, it MUST limit the
```
Comment (ianswett) [inline draft-ietf-quic-transport.md:4104]: I don't think this change is necessary anymore.

**Review by ianswett — COMMENTED** [review]
Some comments, but I agree this is a real issue, which for some reason I thought we already dealt with correctly.

File: draft-ietf-quic-transport.md (hunk 10 above)
Comment (martinduke) [reply draft-ietf-quic-transport.md:4104]: I agree with @martinthomson. The amplification limit disappears when the server validates the address. The padding requirement/recommendation applies to all Initial packets, because you still have to probe the MTU, and the client is often not sure if the server has validated it.

File: draft-ietf-quic-transport.md (hunk 9 above)
Comment (martinduke) [reply draft-ietf-quic-transport.md:1925]: Endpoints that don't coalesce are already trading performance for simplicity. The added sacrifice here isn't that large.

1) Clients: if it only receives an Initial, padding is critical to raising the amplification limit. If it gets a Handshake packet too, it will momentarily send a Handshake ACK that will allow it to free the Initial context and get the cwnd back.

2) Servers: If the client has a valid Retry or NEW_TOKEN token, there is no amplification limit. So the performance penalty only applies to a non-coalescing server when:
(a) There is no token, AND the Initial padding causes the handshake flight to go over the amplification limit.
(b) The initial padding causes the handshake flight to go over the initial cwnd.
I find it hard to believe a server deeply concerned about these losses would not coalesce packets for other reasons.

File: draft-ietf-quic-recovery.md (hunk 6 above)
Comment (ianswett) [reply draft-ietf-quic-recovery.md:667]: I really thought when we discussed this optimization, we said to only do this once, but either I'm misremembering or it never made it into the issue(#3080).

@martinthomson Would the case it's not as good would be repeated loss of the server's Initial packet?

Comment (martinthomson) [reply draft-ietf-quic-recovery.md:667]: Yes, repeated loss is what I'm thinking about.  This is likely rare, but early losses on a flow are more common.  I don't have a firm opinion on this point. One is certainly better than none.  But two doesn't seem to be a real problem either.  For what it is worth, I send up to 4.

Comment (ianswett) [reply draft-ietf-quic-recovery.md:667]: In my experience looking at traces, losing two server Initials in a row is quite rare, and there's still PTO as a backup, so I'm still leaning towards limiting this to once.  Maybe we can require limiting it and specify a RECOMMENDED value of 1 or 2?

Comment (martinthomson) [reply draft-ietf-quic-recovery.md:667]: But now you are telling me what to do without cause :p  But I can live with a recommendation.  My point was that the only hard requirement is that this is limited.

Comment (janaiyengar) [reply draft-ietf-quic-recovery.md:667]: I'm fine with requiring limiting it. To be useful to implementers, we should have some value here, but I would not RECOMMEND it. We can simply say that experience has shown 1 early retransmission to be adequate to accommodate common loss scenarios.

File: draft-ietf-quic-transport.md (hunk 10 above)
Comment (kazuho) [reply draft-ietf-quic-transport.md]: Thank you folks. I think there's consensus that padding should better not be tied to path validation. Addressed the text in 0410b60. PTAL.

File: draft-ietf-quic-recovery.md (hunk 6 above)
Comment (kazuho) [reply draft-ietf-quic-recovery.md:667]: Sounds like we are converging; took a stub in commit 6e0fe53. PTAL.

File: draft-ietf-quic-transport.md [hunk 11]
```diff
@@ -1920,10 +1920,12 @@ that are uniquely attributed to a single connection. This includes datagrams
...
 packets that are all discarded.
 
-Clients MUST ensure that UDP datagrams containing Initial packets have UDP
-payloads of at least 1200 bytes, adding PADDING frames as necessary.
-A client that sends padded datagrams allows the server to
-send more data prior to completing address validation.
+Clients MUST and servers SHOULD ensure that UDP datagrams containing Initial
+packets have UDP payloads of at least 1200 bytes, adding PADDING frames as
+necessary.  A client that sends padded datagrams allows the server to send more
+data prior to completing address validation.  Endpoints using UDP datagrams of
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md:1926]:
```suggestion
data prior to completing address validation.  Endpoints sending UDP datagrams of
```

**Review by martinthomson — APPROVED** [review]
I could tolerate a change to MUST for both client and server, if we added the two tweaks discussed:

* ack-eliciting packets are padded (this is safe for clients too, but less obviously good)
* CONNECTION_CLOSE is not ack-eliciting

I can definitely still live with this.

**Comment by martinthomson** [comment]
FWIW, CONNECTION_CLOSE is already non-ack-eliciting.  I just forgot.

**Review by martinduke — APPROVED** [review]
I can live with these changes, but I would like the CONNECTION_CLOSE text to go in and would prefer a MUST to a SHOULD for servers.

**Comment by kazuho** [comment]
I think nobody is opposing to having a MUST, but the debate is the specifics.

I was hoping if we could converge on the specifics of that MUST, but maybe I should update this PR in one way to narrow down the discussion.

**Comment by martinduke** [comment]
Thank you for clarifying that we agree on a MUST. I understand the concern about validation if we exempt certain packets; however, while regrettable this is no net increase in decrypt operations vs the status quo.

**Comment by kazuho** [comment]
I've updated the pull request so that it describes one specific way of having a MUST. Admittedly, it reflects my preference (the rationale explained in https://github.com/quicwg/base-drafts/issues/4183#issuecomment-708061598), but the hope is that having something specific being written down helps us continue the discussion.

**Review by DavidSchinazi — APPROVED** [review]
Thanks for the discussion, this PR looks great!

**Review by ianswett — APPROVED** [review]
One suggestion on wording, but overall this WFM.

File: draft-ietf-quic-recovery.md [hunk 12]
```diff
@@ -663,9 +663,11 @@ in Initial packets, or the client's estimated RTT is too small. When a
...
+unacknowledged CRYPTO data earlier than the PTO expiry, subject to the address
+validation limits in Section 8.1 of {{QUIC-TRANSPORT}}. Doing so at most once
+for each connection will be adequate to accommodate common loss scenarios.
```
[the remark this answers is not in the document]
Comment (martinthomson) [reply draft-ietf-quic-recovery.md:670]: I don't think that the packet reflection this is well-enough explained to use it here.

[the remark this answers is not in the document]
Comment (kazuho) [reply draft-ietf-quic-recovery.md:670]: I am in favor of @ianswett's suggestion.

The principles we have for preventing ping-pong are:
* do not repeatedly send an ACK-eliciting ACK
* when not receiving an ACK, rely on PTO backoff

This paragraph is talking when an endpoint MAY violate those rules. Therefore, IMO it is good to point out that risk.

[the remark this answers is not in the document]
Comment (martinthomson) [reply draft-ietf-quic-recovery.md:670]: I would rather add:

> A client that does not limit the number of times it sends unacknowledged CRYPTO data in response to unauthenticated data risks creating an infinite exchange of packets.

There are other reasons why this might happen (hitting an endpoint that thinks it closed, for example.)

File: draft-ietf-quic-transport.md [hunk 13]
```diff
@@ -1921,9 +1921,13 @@ that contain packets that are successfully processed and datagrams that contain
...
-payloads of at least 1200 bytes, adding PADDING frames as necessary.
-A client that sends padded datagrams allows the server to
-send more data prior to completing address validation.
+payloads of at least 1200 bytes, adding PADDING frames as necessary.  Servers
+MUST ensure that UDP datagrams containing ack-eliciting Initial packets have UDP
+payloads of at least 1200 bytes.  A client that sends padded datagrams allows
+the server to send more data prior to completing address validation.  Endpoints
+sending UDP datagrams of at least 1200 bytes ensures that the handshake
+progresses only if the path is capable of handling QUIC traffic; see
+{{packet-size}}.
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md:1930]:
```suggestion
payloads of at least 1200 bytes, adding PADDING frames as necessary.  Servers
MUST ensure that UDP datagrams containing ack-eliciting Initial packets have UDP
payloads of at least 1200 bytes. Sending UDP datagrams of at least 1200 bytes
ensures that the handshake progresses only if the path is capable of handling
QUIC traffic; see {{packet-size}}. Additionally, a client that sends padded
datagrams allows the server to send more data prior to completing address
validation.
```

Comment (janaiyengar) [reply draft-ietf-quic-transport.md:1930]: In addition, let's remove the server text from here, since that is unnecessary for address validation. The server datagram padding needs to only be mentioned below, in the Initial Packet Size section.

File: draft-ietf-quic-transport.md [hunk 14]
```diff
@@ -4084,23 +4088,28 @@ fragmentation on the path.
...
-{{packet-coalesce}}.  Sending a UDP datagram of this size ensures that the
-network path from the client to the server supports a reasonable Path Maximum
-Transmission Unit (PMTU).  This also helps reduce the amplitude of amplification
-attacks caused by server responses toward an unverified client address; see
+{{packet-coalesce}}.  Similarly, a server MUST expand the payload of all UDP
+datagrams carrying ack-eliciting Initial packets.  Sending UDP datagrams of this
+size ensures that the network path supports a reasonable Path Maximum
+Transmission Unit (PMTU), in both directions.  Additionally, a client padding
+Initial packets helps reduce the amplitude of amplification attacks caused by
+server responses toward an unverified client address; see
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md:4096]: Can you spell it out fully for the server? "A server MUST similarly expand the payload of all UDP datagrams carrying ack-eliciting Initial packets to at least the smallest allowed maximum packet size (1200 bytes)."

Also, "a client padding Initial packets" --> "a client that expands Initial packets"

File: draft-ietf-quic-transport.md [hunk 15]
```diff
@@ -4084,23 +4088,28 @@ fragmentation on the path.
...
 {{address-validation}}.
 
-Datagrams containing Initial packets MAY exceed 1200 bytes if the client
+Datagrams containing Initial packets MAY exceed 1200 bytes if the endpoint
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md:4099]:
```suggestion
Datagrams containing Initial packets MAY exceed 1200 bytes if the sender
```

File: draft-ietf-quic-transport.md [hunk 16]
```diff
@@ -4084,23 +4088,28 @@ fragmentation on the path.
...
 CONNECTION_CLOSE frame with an error code of PROTOCOL_VIOLATION; see
-{{immediate-close-hs}}.
+{{immediate-close-hs}}.  When a client receives an ack-eliciting Initial packet
+that is carried in a UDP datagram with a payload that is less than 1200 bytes,
+that client MAY close the connection by sending a CONNECTION_CLOSE frame.
 
 The server MUST also limit the number of bytes it sends before validating the
-address of the client; see {{address-validation}}.
+address of the client, and MUST pad ack-eliciting Initial packets that it sends;
+see {{address-validation}}.
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md:4112]:
```suggestion
address of the client; see {{address-validation}}.
```

I don't think you need this here, since this is described earlier in this very section.

**Review by janaiyengar — COMMENTED** [review]
Thanks, @kazuho. I can live with this. A few comments, editorial.

File: draft-ietf-quic-recovery.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (kazuho) [reply draft-ietf-quic-recovery.md]: In 3518d54, applied changes suggested by @ianswett to the existing sentence.
In a51ad56, added the rationale suggested by @martinthomson with tweaks. Specifically, I changed the subject to "endpoints" so that it would be generic, and used "back off", as we suggest using that for Stateless Resets, etc.

PTAL.

File: draft-ietf-quic-transport.md (hunk 13 above)
Comment (kazuho) [reply draft-ietf-quic-transport.md]: @janaiyengar Following your last comment, I've reverted the changes to this section (i.e. address validation). That means that the suggested change right above is not applied. Please see if you want anything to be resurrected.

File: draft-ietf-quic-transport.md (hunk 14 above)
Comment (kazuho) [reply draft-ietf-quic-transport.md]: Thank you for the suggestions. Done in c550cb4.

File: draft-ietf-quic-recovery.md [hunk 17]
```diff
@@ -663,9 +663,13 @@ in Initial packets, or the client's estimated RTT is too small. When a
...
 
-To speed up handshake completion under these conditions, an endpoint MAY send
-a packet containing unacknowledged CRYPTO data earlier than the PTO expiry,
-subject to the address validation limits in Section 8.1 of {{QUIC-TRANSPORT}}.
+To speed up handshake completion under these conditions, an endpoint MAY, for a
+limited number of occasions per each connection, send a packet containing
+unacknowledged CRYPTO data earlier than the PTO expiry, subject to the address
+validation limits in Section 8.1 of {{QUIC-TRANSPORT}}. Doing so at most once
+for each connection is adequate to quickly recover from a single packet loss.
+Endpoints that do not back off from retransmitting packets in response to
```
Comment (martinthomson) [inline draft-ietf-quic-recovery.md:671]:
```suggestion
Endpoints that do not cease retransmitting packets in response to
```

File: draft-ietf-quic-recovery.md [hunk 18]
```diff
@@ -663,9 +663,13 @@ in Initial packets, or the client's estimated RTT is too small. When a
...
+unauthenticated data risks creating an infinite exchange of packets.
```
Comment (martinthomson) [inline draft-ietf-quic-recovery.md:672]:
```suggestion
unauthenticated data risk creating an infinite exchange of packets.
```

**Review by martinthomson — APPROVED** [review]
Tweaks.

**Comment by martinthomson** [comment]
Thanks for patiently working through this @kazuho.  This was difficult, but the fix was, in the end, fairly tidy.

**Comment by janaiyengar** [comment]
Apologies for leaving the "Requires Changes" flag on. This PR LGTM.

**Comment by marten-seemann** [comment]
Why was this already merged? I don't believe that this PR is a good solution for the issue, as I've detailed in https://github.com/quicwg/base-drafts/issues/4183#issuecomment-708876280. 

I still think we should include ACK-only packets in the padding requirement. This would allow us to get rid of the change made in this PR that limits the number of retransmissions on undecryptable packets, and it would prevent a ping-pong (even if that ping-pong comes with an exponential backoff).
